File: Assignment5/VB.NET/AritMeanDistr/AritMeanDistr/Distribuzione.vb
Imports System
Imports System.Collections.Generic
Imports System.Runtime.InteropServices

Namespace WindowsFormsApp1
    Friend Class Distribuzione
        Private distr As Dictionary(Of String, SortedDictionary(Of Tuple(Of Double, Double), Integer))
        Private intervall As Double = 10

        Public Sub New()
            distr = New Dictionary(Of String, SortedDictionary(Of Tuple(Of Double, Double), Integer))()
        End Sub

        'intervall standard 10
        Public Sub addAttributDef(ByVal s As String, ByVal i As Double)
            addAttribute(s, i, intervall)
        End Sub

        Public Sub addAttribute(ByVal s As String, ByVal value As Double, ByVal inte As Double)
            Dim actualdistr As SortedDictionary(Of Tuple(Of Double, Double), Integer)
            Dim min, max As Double
            Dim i As Integer = 1
            min = value - value Mod inte
            max = value + (inte - value Mod inte)
            Dim tmp As Tuple(Of Double, Double) = New Tuple(Of Double, Double)(min, max)

            If Not distr.TryGetValue(s, actualdistr) Then
                actualdistr = New SortedDictionary(Of Tuple(Of Double, Double), Integer)()
                actualdistr.Add(tmp, 1)
                distr.Add(s, actualdistr)
            Else

                If Not actualdistr.TryGetValue(tmp, i) Then
                    actualdistr.Add(tmp, 1)
                Else
                    i += 1
                    actualdistr.Remove(tmp)
                    actualdistr.Add(tmp, i)
                End If

                distr.Remove(s)
                distr.Add(s, actualdistr)
            End If
        End Sub

        Public Sub addElemento(ByVal e As ElementoDisribuzione, ByVal inter As Double)
            For Each item In e.getVariabili()
                addAttribute(item.Key, item.Value, inter)
            Next
        End Sub

        Public Sub addElementoDef(ByVal e As ElementoDisribuzione)
            Me.addElemento(e, intervall)
        End Sub

        Public Function getdistribuzione(ByVal s As String, <Out> ByRef req As SortedDictionary(Of Tuple(Of Double, Double), Integer)) As Boolean
            If distr.TryGetValue(s, req) Then
                Return True
            Else
                Return False
            End If
        End Function
    End Class
End Namespace

File: Assignment5/C#/OnlineArithmeticMean/WindowsFormsApp1/Distribuzione.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace WindowsFormsApp1
{
    class Distribuzione

    {
        private Dictionary<String, SortedDictionary<Tuple<double, double>, int>> distr;
        double intervall = 10;

        public Distribuzione()
        {
            distr = new Dictionary<String, SortedDictionary<Tuple<double, double>, int>>();

        }
        //intervall standard 10
        public void addAttributDef(String s, double i)
        {
            addAttribute(s, i, intervall);
        }
        public void addAttribute(String s, double value, double inte)
        {
            SortedDictionary<Tuple<double, double>, int> actualdistr;
            double min, max;
            int i = 1;
            min = value - (value % inte);
            max = value + (inte - (value % inte));
            Tuple<double, double> tmp = new Tuple<double, double>(min, max);

            if (!distr.TryGetValue(s, out actualdistr))
            {
                actualdistr = new SortedDictionary<Tuple<double, double>, int>();
                actualdistr.Add(tmp, 1);
                distr.Add(s, actualdistr);

            }
            else
            {
                if (!actualdistr.TryGetValue(tmp, out i)) actualdistr.Add(tmp, 1); 
                else { i++; actualdistr.Remove(tmp); actualdistr.Add(tmp, i); }
                distr.Remove(s); 
                distr.Add(s, actualdistr);
            }
        }

        public void addElemento(ElementoDisribuzione e, double inter)
        {
            foreach (var item in e.getVariabili()) {
                this.addAttribute(item.Key, item.Value, inter);
            }


        }
        public void addElementoDef(ElementoDisribuzione e)
        {
            addElemento(e, intervall);
        }



        public bool getdistribuzione(string s, out SortedDictionary<Tuple<double, double>, int> req)
        {
     
            if (distr.TryGetValue(s, out req)) return true;
            else return false;

        }
    }  
    }

